### PR TITLE
Block Creation Simplification

### PR DESCRIPTION
**Pull request by ymte**
#83

# Introduction
BlockController has obtained new methods that do a lot of heavy lifting around the creation and adding of new blocks.

# Purpose
It becomes easier and less painful to add blocks now.

# Future Applicability
The BlockFactory class will be obsolete and should be filtered out.

# PR Reflection
There are no complications currently since the code is not actively used; it will be once BlockFactory is gone.

### CODE REVIEW

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java [hunk 1]
```diff
@@ -228,4 +228,81 @@ public final boolean isDatabaseEmpty() {
...
+            contactBlockHash = "N/A";
+        else
+            contactBlockHash = getBlocks(contact).get(0).getOwnHash();
+        int index = latest.getSequenceNumber() + 1;
+
+        ConversionController conversionController = new ConversionController(
+                owner, publicKey, previousBlockHash, contactBlockHash, iban
+        );
+        String hash = conversionController.hashKey();
+        Block block = new Block(index, owner, hash, previousBlockHash, contactBlockHash, publicKey, iban, 0, revoke);
```
Comment (ashay123) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Instead of setting the trustValue to 0 in a hardcoded way, make in the TrustValue class, another enum Create_Block, or Initialize_Block

Comment (ymte) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: This must be done in another pull request, using a fresh TrustValues item that does not yet exist.

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java [hunk 2]
```diff
@@ -228,4 +228,81 @@ public final boolean isDatabaseEmpty() {
...
+     * @throws Exception when the key hashing method does not work
+     */
+    public Block createGenesis(String owner) throws Exception {
+        String chainHash = "N/A";
+        String senderHash = "N/A";
+        String publicKey = "N/A";
+        String iban = "N/A";
+        ConversionController conversionController = new ConversionController(owner, publicKey, chainHash, senderHash, iban);
+        String hash = conversionController.hashKey();
+        Block block = new Block(1, owner, hash, chainHash, senderHash, publicKey, iban, 0, false);
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Not the way to create a block, use the blockfactory to generate a block.

Comment (ymte) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Done.

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java [hunk 3]
```diff
@@ -228,4 +228,83 @@ public final boolean isDatabaseEmpty() {
...
+     * @throws Exception when the key hashing method does not work
+     */
+    public Block createGenesis(String owner) throws Exception {
+        String chainHash = "N/A";
+        String senderHash = "N/A";
+        String publicKey = "N/A";
+        String iban = "N/A";
+        ConversionController conversionController = new ConversionController(owner, publicKey, chainHash, senderHash, iban);
+        String hash = conversionController.hashKey();
+        Block block = BlockFactory.getBlock("BLOCK", owner, hash, chainHash, senderHash, publicKey, iban, 0);
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Split in multiple lines

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java [hunk 4]
```diff
@@ -228,4 +228,83 @@ public final boolean isDatabaseEmpty() {
...
+     * @param owner the new owner of the block
+     * @return the freshly created block
+     * @throws Exception when the key hashing method does not work
+     */
+    public Block createGenesis(String owner) throws Exception {
+        String chainHash = "N/A";
+        String senderHash = "N/A";
+        String publicKey = "N/A";
+        String iban = "N/A";
+        ConversionController conversionController = new ConversionController(owner, publicKey, chainHash, senderHash, iban);
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Split in multiple lines

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java [hunk 5]
```diff
@@ -244,6 +264,71 @@ public void testRevokedTrustValue() {
...
+
+    /** Verify block hash */
+    @Test
+    public void verifySenderHash() {
+        try {
+            ConversionController conversionController = new ConversionController("A", "kb", blockA_A1.getOwnHash(), genesisB.getOwnHash(), "ibanB");
+            String hash = conversionController.hashKey();
+            assertEquals(hash, blockA_B1.getOwnHash());
+        } catch (Exception e) {
+            assertNotNull(null);
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: ?

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java [hunk 6]
```diff
@@ -209,4 +209,83 @@ public final boolean isDatabaseEmpty() {
...
+            contactBlockHash = "N/A";
+        else
+            contactBlockHash = getBlocks(contact).get(0).getOwnHash();
+        int seqNumber = latest.getSequenceNumber() + 1;
+
+        ConversionController conversionController = new ConversionController(
+                owner, publicKey, previousBlockHash, contactBlockHash, iban
+        );
+        String hash = conversionController.hashKey();
+        Block block = BlockFactory.getBlock(revoke?"REVOKE":"BLOCK", owner, hash, previousBlockHash, contactBlockHash, publicKey, iban, 0);
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: multiple lines, spaces

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java [hunk 7]
```diff
@@ -209,4 +209,83 @@ public final boolean isDatabaseEmpty() {
...
+
+        // always link to genesis of contact blocks
+        String contactBlockHash;
+        if (owner.equals(contact))
+            contactBlockHash = "N/A";
+        else
+            contactBlockHash = getBlocks(contact).get(0).getOwnHash();
+        int seqNumber = latest.getSequenceNumber() + 1;
+
+        ConversionController conversionController = new ConversionController(
```
Comment (humanjas) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: multiple lines, spaces

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java [hunk 8]
```diff
@@ -242,6 +262,67 @@ public void testRevokedTrustValue() {
...
+    /** Verify genesis owner A */
+    @Test
+    public void verifyGenesisOwner() {
+        assertEquals("A", genesisA.getOwner());
+    }
+
+    /** Hash of genesis block */
+    @Test
+    public void verifyGenesisHashA() {
+        ConversionController conversionController = new ConversionController("A", "N/A", "N/A", "N/A", "N/A");
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: create "N/A" attribute

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java [hunk 9]
```diff
@@ -242,6 +262,67 @@ public void testRevokedTrustValue() {
...
+    /** Blocks without sender should not have a sender hash */
+    @Test
+    public void verifyBlockChainHash() {
+        assertEquals("N/A", blockA_A1.getPreviousHashSender());
+    }
+
+    /** Verify block hash */
+    @Test
+    public void verifySenderHash() throws Exception {
+        ConversionController conversionController = new ConversionController("A", "kb", blockA_A1.getOwnHash(), genesisB.getOwnHash(), "ibanB");
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: multiple lines

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java [hunk 10]
```diff
@@ -242,6 +262,67 @@ public void testRevokedTrustValue() {
...
+        String hash = conversionController.hashKey();
```
Comment (humanjas) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: make final

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java [hunk 11]
```diff
@@ -242,6 +263,69 @@ public void testRevokedTrustValue() {
...
+            String hash = conversionController.hashKey();
+            assertEquals(genesisA.getOwnHash(), hash);
+        } catch (Exception e) {
+            assertNotNull(e.getMessage(), null);
+        }
+    }
+
+    /** Verify the block A-A1 was created */
+    @Test
+    public void verifyBlockA_A1Creation() {
```
Comment (nzarin) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: What is A_A1? Can you change this and be more specifiek. The attribute names are pretty important.

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java (hunk 7 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Fixed

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java (hunk 11 above)
Comment (nzarin) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: What is A_A1? Can you change this and be more specifiek. The attribute names are pretty important.

Comment (ymte) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: Agreed upon. Done.

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java (hunk 5 above)
Comment (ymte) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: fixed

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java (hunk 10 above)
Comment (ymte) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: Done.

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java (hunk 3 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Done.

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java (hunk 4 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Done.

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java (hunk 9 above)
Comment (ymte) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: Done.

File: app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java (hunk 8 above)
Comment (ymte) [inline app/src/test/java/nl/tudelft/b_b_w/controller/BlockControllerUnitTest.java]: Done.

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java (hunk 6 above)
Comment (ymte) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Done.

File: app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java (hunk 1 above)
Comment (ashay123) [inline app/src/main/java/nl/tudelft/b_b_w/controller/BlockController.java]: Alright.
